### ISSUE TITLE
aes ocb bug

### ISSUE THREAD

**Issue by nrgiii** [openssl/openssl#2273]
This is for OpenSSL 1.1.0c

Encrypting or decrypting using AES OCB is broken if the input data size is not a multiple of the block size (16 bytes).  The problem is in crypto/evp/e_aes.c in the function aes_ocb_cipher in this block:
```
2560         if (*buf_len) {
2561             unsigned int remaining;
2562
2563             remaining = 16 - (*buf_len);
2564             if (remaining > len) {
2565                 memcpy(buf + (*buf_len), in, len);
2566                 *(buf_len) += len;
2567                 return 0;
2568             }
2569             memcpy(buf + (*buf_len), in, remaining);
2570
2571             /*
2572              * If we get here we've filled the buffer, so process it
2573              */
2574             len -= remaining;
2575             in += remaining;
2576             if (out == NULL) {
2577                 if (!CRYPTO_ocb128_aad(&octx->ocb, buf, 16))
2578                     return -1;
2579             } else if (EVP_CIPHER_CTX_encrypting(ctx)) {
2580                 if (!CRYPTO_ocb128_encrypt(&octx->ocb, buf, out, 16))
2581                     return -1;
2582             } else {
2583                 if (!CRYPTO_ocb128_decrypt(&octx->ocb, buf, out, 16))
2584                     return -1;
2585             }
2586             written_len = 16;
2587             *buf_len = 0;
2588             out += 16; /*  <=== Missing line of code */
2589         }
```

Here we are building a 16 byte block by combining previously supplied data which was less than 16 bytes with new input data.  The problem is that we do not advance the output pointer, which leads to the data being overwritten.

Adding line 2588 fixes the bug.

Seen when encrypting data with the EVP_ase_256_ocb cipher when one or more input data block sizes are not a multiple of 16 and at least one said data block is NOT the last block.

Norm Green

**Comment by mattcaswell** [comment]
The fix doesn't seem quite right. `out` could be NULL here (in the case of AAD data).

**Comment by levitte** [comment]
The fix should depend on if `out` is non-NULL, and seems to need to be added in more spots.
I can propose a patch, but someone else will have to come up with an appropriate test.  AAD isn't my forte...

**Comment by levitte** [comment]
I was wrong, it's only in this spot that it's needed.  It still does need to be made conditional, though.

**Comment by mattcaswell** [comment]
Fix in #2275
